ray-curator/ray_curator/stages/classifiers/aegis.py

```diff
@@ -0,0 +1,555 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+os.environ["RAPIDS_NO_INITIALIZE"] = "1"
+from dataclasses import dataclass
+from typing import Literal
+
+import numpy as np
+import pandas as pd
+import torch
+import torch.nn.functional as F  # noqa: N812
+from huggingface_hub import PyTorchModelHubMixin
+from torch import nn
+from torch.nn import Dropout, Linear
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from ray_curator.backends.base import WorkerMetadata
+from ray_curator.stages.base import CompositeStage, ProcessingStage
+from ray_curator.stages.classifiers.base import HFModel, HFTokenizerStage
+from ray_curator.stages.modules.score_filter import Filter
+from ray_curator.tasks import DocumentBatch
+
+from .aegis_utils import AEGIS_LABELS, format_aegis
+
+
+class InstructionDataGuardNet(torch.nn.Module, PyTorchModelHubMixin):
+    def __init__(self, input_dim: int, dropout: float = 0.7):
+        super().__init__()
+        self.input_dim = input_dim
+        self.dropout = Dropout(dropout)
+        self.sigmoid = torch.nn.Sigmoid()
+        self.input_layer = Linear(input_dim, input_dim)
+
+        self.hidden_layer_0 = Linear(input_dim, 2000)
+        self.hidden_layer_1 = Linear(2000, 500)
+        self.hidden_layer_2 = Linear(500, 1)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = F.normalize(x, dim=-1)
+        x = self.dropout(x)
+        x = F.relu(self.input_layer(x))
+        x = self.dropout(x)
+        x = F.relu(self.hidden_layer_0(x))
+        x = self.dropout(x)
+        x = F.relu(self.hidden_layer_1(x))
+        x = self.dropout(x)
+        x = self.hidden_layer_2(x)
+        return self.sigmoid(x)
+
+
+class AegisModel(nn.Module):
+    def __init__(  # noqa: PLR0913
+        self,
+        pretrained_model_name_or_path: str,
+        peft_model_name_or_path: str,
+        dtype: torch.dtype,
+        token: str | bool | None,
+        add_instruction_data_guard: bool = False,
+        autocast: bool = False,
+    ):
+        super().__init__()
+        base_model = AutoModelForCausalLM.from_pretrained(
+            pretrained_model_name_or_path, torch_dtype=dtype, token=token
+        )
+        # Importing PeftModel here to prevent cuda context issues
+        # that seem to happen on Transformers 4.48.3
+        # See related: https://github.com/rapidsai/crossfit/pull/113
+        from peft import PeftModel
+
+        self.model = PeftModel.from_pretrained(base_model, peft_model_name_or_path)
```

Comment:
shouldn't this be in setup?

Reply:
I can double check to see if the model can be moved elsewhere. Main thing to avoid here are the CUDA issues.